# Puzzle 1

--- Day 1: Not Quite Lisp ---

Papá Noel esperaba una Navidad blanca, pero la función de "nieve" de su máquina meteorológica funciona con estrellas, ¡y acaba de salir! Para salvar la Navidad, necesita que recolectes cincuenta estrellas antes del 25 de diciembre.

Recoge estrellas ayudando a Santa a resolver acertijos. Dos rompecabezas estarán disponibles cada día en el calendario de Adviento; el segundo rompecabezas se desbloquea cuando completas el primero. Cada rompecabezas otorga una estrella. ¡Buena suerte!

Aquí tienes un sencillo rompecabezas para que entres en calor.

Santa está tratando de entregar regalos en un gran edificio de apartamentos, pero no puede encontrar el piso correcto; las instrucciones que obtuvo son un poco confusas. Comienza en la planta baja (piso 0) y luego sigue las instrucciones un carácter a la vez.

Un paréntesis de apertura, (, significa que debe subir un piso, y un paréntesis de cierre,), significa que debe bajar un piso.

El edificio de apartamentos es muy alto y el sótano es muy profundo; nunca encontrará los pisos superior o inferior.

Por ejemplo:

(()) y () () ambos dan como resultado el piso 0.
(((y (() (() (ambos resultan en el piso 3.
)) ((((((también resulta en el piso 3.
()) y)) (ambos resultan en el piso -1 (el primer nivel del sótano).
))) y) ()) ()) ambos dan como resultado el piso -3.
¿A qué piso llevan las instrucciones a Santa?


## Logica de Resolucion

Como vemos la entrada va a ser una cadena donde encontraremos dentro muchos caracteres que pueden ser un parentesis abierto '(' o un parentesis cerrado ')' y tendremos que utilizar estos para decifrar a que piso debe llegar Santa, para ello:

* asignaremos a la variable `mensaje` el contenido de nuestro `input`

* Vamos a tener una variable llamada `piso` que va a guardar el valor del piso actual.

* Con un `ciclo for` vamos a ir recorriendo la palabra y viendo cada uno de los caracteres

* Ahora tomaremos una `decision`:
    
    * Si el caracter es '(' la variable `piso` incrementara su valor en 1
    
    * Si el caracter es ')' la variable `piso` decrementara su valor en 1


* Al final imprimiremos la variable `piso` que nos dara el va

## Mis Datos

Los datos que me dieron para el ejercicio si quieres practicar lo puedes encontrar en el archivo input_ejercicio1.txt

## Codigo

In [2]:
contenido = open('input_puzzle1.txt', 'r')
mensaje = contenido.read()
piso = 0
for letra in mensaje:
    if letra == '(':
        piso += 1
    if letra == ')':
        piso -= 1

print(piso)

232


## Resultado de Parte 1: 232

## Parte 2

--- La segunda parte ---
Ahora, dadas las mismas instrucciones, encuentra la posición del primer personaje que le hace entrar al sótano (piso -1). El primer carácter de las instrucciones tiene la posición 1, el segundo carácter tiene la posición 2 y así sucesivamente.

Por ejemplo:

) hace que entre al sótano en la posición de personaje 1.
() ()) hace que entre al sótano en la posición de personaje 5.
¿Cuál es la posición del personaje que hace que Santa ingrese por primera vez al sótano?


## Logica de Resolucion

Como vemos ahora lo que nos interesa es saber el indice en el que se encuentra el primer 'personaje' que hace que santa baje al piso -1, para eso al ejercicio anterior le vamos a agregar:

* En una variable `contador` vamos a asignar el valor del indice en el que se encuentre el personaje actual, entonces ira incrementando en cada vuelta del ciclo, ¡OJO! que debe comenzar en `1` (por la consigna).
* Como recordamos antes en piso guardabamos el valor del piso, entonces iremos vienddo que el valor de `piso` sea `-1` y ahi utilizamos un break para salir del ciclo y que quede guardado en `contador` el valor de ese personaje
* Mostraremos el valor de `contador` que tendra el valor de personaje

## Codigo

In [2]:
mensaje = '(((())))()((((((((())()(()))(()((((()(()(((()((()((()(()()()()()))(((()(()((((((((((())(()()((())()(((())))()(()(()((()(()))(()()()()((()((()(((()()(((((((()()())()((((()()(((((()(())()(())((())()()))()(((((((())(()())(()(((())(()))((())))(()((()())))()())((((())))(()(((((()(())(((()()((()((()((((((((((())(()())))))()))())()()((((()()()()()()((((((())())(((()())()((()()(((()()()))(((((()))(((()(()()()(()(()(((())()))(()(((()((())()(()())())))((()()()(()()(((()))(((()((((()(((((()()(()())((()())())(()((((((()(()()))((((()))))())((())()()((()(()))))((((((((()))(()()(((())())(())()((()()()()((()((()((()()(((())))(()((())()((((((((()((()(()()(((())())())))(())())))()((((()))))))())))()()))()())((()())()((()()()))(()()(((()(())((((())())((((((((()()()()())))()()()((((()()))))))()((((()(((()))(()()())))((()()(((()))()()())())(((())((()()(())()()()(((())))))()())((()))()))((())()()())()())()()(()))())))())()))(())((()(())))(()(())(()))))(()(())())(()(())(()(()))))((()())()))()((((()()))))())))()()())((())()((()()()))()(((()(()))))(())()()))(((()())))))))))(((())))()))())()))))()()(((())))))))()(()()(()))((()))))((())))((()((())))())))()()(()))())()(()((()())(()(()()())())(()()))()))))(()())()()))()()()()))(()(()(()))))))()(()))()))()()(()((())(()(())))()(((())(())())))))()(()(()))))()))(()()()(())()(()(())))()))))()()(((((())))))())()())())())()())()))))()))))))))())()()()()()()())))()))((())()))())))()((())()))))()))())))))))())()()()))()()(()((((()(((((((()(())((()())((()()))()))))(())))()()()(())((())()())))(())))(())))(((()()))()(())(((()(()))((())))())()))((((()))())()))))))))()(())())))(()))()(()()))())()()(())())))())()()(()())))()((()())(()(())(())))))))))))))(()))))()))))))()()())(()(((((()(()())))())()))(()))()))(()()))()())(()))())()(())((()()))))))())))())()(((())))(()(()))()()))()(()))))))((()())(()))))))()())))()()))))))))((((((((()()()(()))))))()())))())))()()((())()))((())(())))())())))()()()((()((()(())))())()(())))))))))()())))()()()()()()))()))((())())(()(()))))))(()()))()))(())))()))))))))))))(()))))))))()))))()))()())()))()()))))))()))))((()))))(()))())()(())))(()())((((()())))()))))(()))()(()()(())))))())))))()))))))())))())))))())))())())))())(()))))(())()(())))())()))((()()))))))())))((())))))))())))(())))))()()())))))())))))()))))))()))()()()(()(((()())())())(()))())))))((()(())(()))))))))(())))()()()())())(()))))()()()))()))())())())()(())))()(((()((((())))))))()))))))))))))))))))))((())()())(()))))()()))))))(()()(())())))())))((())))((())))))))))))))()))))()(()))))))())))))()))(()()())(()())))))))))()))))))(())))))()()))()())(((())))()))(()))))))))(())())))())))())())())()()))((())()(())()())()))()())(())(()))))()())))(()(((()))))))()(()())()()()))()))))))))()()()(())()())()(((((()))()())())(()))))()()()(())))())))()((()())))(()))())()(()())())(()))()()))((()()))((()()()()())))(())()))(()(())))((()()))))))))())))))))())()()))))))))))))))))(())()(())(())()())())()))()(()))))())())))))()())()(()))()()(())))(())())))))(()))))))))))))))())())(())(())))(((()))()))))())((())(()))())))))))())))))())))()))()))))))))))))())()))))()))))((()))(())))()(())))(())()))()))())))())))))))()(()())())))()()())))(())))))(()))))))))))))(()))()))()))())))(((()()()(())((()())))()())(((()))(())()))((()()()())))())(())(()))))()(((((())))(()))())())))))))((((()()()))())())()(()(()())))))))))()())())))(())))()())(((()(())())()()))())())))))))((()())((()()(()))(()(())))()))()))(()))(()))()()(()(((())((((()))()(()))((())()(()(()())()(()))()())))))(()))()))())()())))())))(())))((())(()())))))()))(())(()))()())()(()()((()(()))))))()(())(()())(())()))(((())()))(()()(()()()))))(()(())))()))))())))))())(()()()()()()(((())))(()()))()((())(((((()()())))(()))(()))()()))(((())())()(((()()()()))))(()))(())())))()())(()()())())))))))()))))((())))()())(()))(()(()))())))))())(())))))()()())())()))()()(())))(()))(())((((((())(()))(()))())()))(()()(())))()))(()()))()))()(())))(())))((()(()))(())()()())())))(((()()())(())()))))))()(((()(((((()()(((())(())))())()((()))))((()())()(())(((())))(((()((()(()(()))(()()))())(()))(())(())))()))))))((((()))()((((()(()))()))()()))))()(()(()))()(()((()(((()(()()(((()))))()(((()(()(()(((()(()())())()()(()(()())())(()((((())(()))()))(((((()()())(())()((()()())))()()(((()()))()((((((((()(())))())((()))))(())))(()))))((()((((()()(())(((((()))(((((((((((((()())))((((()(((()((())())()))((()))()(()()((()()()()(()()(()(()(((())()(()((((((()((()()((())()((((()((()()(()()())((()()()((()((())()(()(((()((())((((())(()))((()(()))(()())()((((((((()(((((((((((()))(()(((()(()()()((((())((())()())()))(())((())(()))(((()((()(())))(()))))((()()))))((((()(()(()())(()(())((((((((()((((()((()(((((()))())()(()))(()()((()(())(((((()(())()(((((()()))))))()(((())()(()()((((())()((())((()(((())(((()))((()()((((()(())))))((()((((()((()((()(((())((()))(((((((()(((()((((((((())()))((((())(((((()((((((((()(((()((()(((()()(((()((((((()()(()((((((((()()(()(()(())((((()())()))))(((()))((((())((((()())((()(())()((()((((((()((((((()(())))()())(((())())())()(())()(()())((()()((((())((((((())(()(((((()((((())()((((()(()(())(()())(((())()((())((((()))()((((((())(()(((()(((()((((((()(((()))(()()())())((()((()())()((((())(((()(()(((((((((())(())))()((()()()()(())((()))(((((((()(((((((((()(()))))(()((((((((()((((()((()()((((((()()(((((((()(()(())()(())((()()()((()(((((()())()(((((()())()()((()(()())(()()()(((()()(((((()((((((()()((()(()()()((((((((((((()((((((((()()(((()())))()(((()()(())())((((()((((()((((()()()(())(())((()(()(((((((((((((((()(())(())))))()()))((()(((()(())((()(((()(()()((((()()(((()(((()(((((()()((()(()(((()))((((((()((((((((()((()((())(((((()(((())(())())((()()))((((())()()((()(((()(((((()()(((()))(((()(()(((((((((((((()))((((((((()(((()))))())((((((((((((())((())((()())(((())((())(()((((((((((()(((())((()()(()((())(((((((((((()))((((((((((((()(()())((()((()((()(()(((()((((((((()()(()((()(()(((()))((()))(((((((((((((()(())((((((())(((()(())(()(()(()((()()))((((()((((()((((())))())((((()((((()))((((((()((((((()((()(((())))((())(()))(()((()((((()((()(((()()))((((()()()(((((((())(((())(()))())((((()())(((()(((((((((((()(()(()((()(((((((((((((((()()((((()((((((((()(((()()((()((((()))(((()(())((((((()((((())()((((()((()))(())()(()(((()((())())((((((()(()(())())(((())(()(()())(((((()((()((())()())(())))(((()(())))))))(((()(((()))()((()(((()()((()())()()))())))(((()))(()(((()(((((((((()(()(((((()()(((()())()()))))()(((()))(((()(()(()(()(()))()(())()))(()(((())))(()))))))))))(())((()((())((()(())()(())((()()((((()()((()()))((())(((()((()(())(())))()(()(((((()((()))())()(((((()()(((()(()((((((())(()))(())()))((()(()()))(())())()))(((())))(()((()(((())(())())))((()()((((((((((((((()((()(()()(()(((()))())()()((()()()(())(()))(()())(((())((())()(())()()(()()(())))((()(((()))))(((()()(()()))())((()((())()))((((()()()())((())))(((()(())(((((()(((((()((()(()((((()()(((()()()(((()())(((()()((((())(()))(((()))(())())((()))(((()((()))(((()()((())((()(((((()((((()()())((()))()((((()((()(()()()('
piso = 0
contador = 1 # agregado
for letra in mensaje:
    if letra == '(':
        piso += 1
    if letra == ')':
        piso -= 1
    if piso == -1: # agregado
        break # agregado
    contador += 1 # agregado
print(contador) # agregado

1783


## Resultado de Parte 2: 1783